File: matlab/sb.m
fprintf("\n%%%%%% problema 1\n");
a = bisection(-4, -3, 0.01, 1);
fprintf("\n%%%%%% problema 2\n");
b = bisection(1, 2, 0.01, 1);

function output = caso1(x)
    output = x .^ 4 - 9 * x .^ 3 - 2 * x .^ 2 + 120 * x - 130;
end

function root = bisection(x1, x2, err, count)
    if count == 1
        fprintf(">> f = inline('x.^4-9*x.^3-2*x.^2+120*x-130')\n");
        fprintf("\nf =\n\n\tInline function:\n\tf(x) = x.^4-9*x.^3-2*x.^2+120*x-130\n\n")
    end

    fprintf("%%%% iteracion %d\n", count);
    fprintf(">> x1 = %g, x2 = %g;\n", x1, x2);

    xm = (x1 + x2) / 2;
    fprintf(">> xm = (x1 + x2) / 2\n");
    fprintf("\nxm =\n\n\t%g\n\n", xm);

    delta = abs(x2 - x1);
    fprintf(">> abs(x2 - x1)\n");
    fprintf("\nans =\n\n\t%g\n\n", delta);
    if delta <= err
        fprintf("%% %g <= %g, alfa = %g\n", delta, err, xm);
        fprintf("%% comprobacion\n");
        fprintf(">> f(xm)\n")
        fprintf("\nans =\n\n\t%g\n\n", caso1(xm));
        root = xm;
        return
    end

    f_x1 = caso1(x1);
    fprintf(">> f(x1)\n");
    fprintf("\nans =\n\n\t%g\n\n", f_x1);
    f_xm = caso1(xm);
    fprintf(">> f(xm)\n");
    fprintf("\nans =\n\n\t%g\n\n", f_xm);

    product = f_x1 * f_xm;
    fprintf(">> %g * %g\n", f_x1, f_xm);
    fprintf("\nans =\n\n\t%g\n\n", product);
    fprintf("%% %g <= 0 ? ", product);
    if product <= 0
        fprintf("T\n");
        fprintf("%% alfa esta en el intervalo (%g, %g)\n", x1, xm);
        root = bisection(x1, xm, err, count + 1);
    else
        fprintf("F\n");
        fprintf("%% alfa esta en el intervalo (%g, %g)\n", xm, x2);
        root = bisection(xm, x2, err, count + 1);
    end
end

File: python/sb.py
from math import log, ceil
from ecuaciones import caso1

def countSteps(d, error):
    calc = (log(d) - log(error)) / log(2)
    return ceil(calc)

def bisection(x1, x2, count = 1, error = 0.01, f = caso1):
    delta = abs(x2 - x1)
    if count == 1:
        print("This process will take", countSteps(delta, error), "iterations.")
    print("\n----------", count, "----------")
    print("x1 =", x1, "      x2 =", x2)

    xm = (x1 + x2) / 2
    print("xm =", xm)
    print("|x2 - x1| =", delta)

    if delta <= error:
        print("alpha = xm =", xm)
        return xm

    product = f(x1) * f(xm)
    print("f(x1) = f(", x1, ") =", f(x1))
    print("f(xm) = f(", xm, ") =", f(xm))
    print("f(x1)f(xm) =", product)
    if product <= 0:
        print("alpha -> (", x1, ",", xm, ")")
        print("x2 = xm =", xm)
        return bisection(x1, xm, count + 1)
    else:
        print("alpha -> (", xm, ",", x2, ")")
        print("x1 = xm =", xm)
        return bisection(xm, x2, count + 1)